src/app/3d/qgs3dmaptoolmeasureline.cpp

```diff
@@ -0,0 +1,269 @@
+
+    // Add feature to layer
+    mMeasurementLayer->startEditing();
+    mMeasurementLayer->addFeature( *measurementFeature );
+    mMeasurementLayer->commitChanges();
+
+    // Set style
+    updateSettings();
+    mIsAlreadyActivated = true;
+//    updateSettings();
+  }
+
+  // Show dialog
+  mDialog->updateSettings();
+  mDialog->show();
+}
+
+void Qgs3DMapToolMeasureLine::deactivate()
+{
+  restart();
+  if ( QgsTerrainEntity *terrainEntity = mCanvas->scene()->terrainEntity() )
+  {
+    disconnect( terrainEntity->terrainPicker(), &Qt3DRender::QObjectPicker::clicked, this, &Qgs3DMapToolMeasureLine::onTerrainPicked );
+  }
+
+  mCanvas->scene()->unregisterPickHandler( mPickHandler.get() );
+
+  // Hide dialog
+  mDialog->hide();
+}
+
+void Qgs3DMapToolMeasureLine::onTerrainPicked( Qt3DRender::QPickEvent *event )
+{
+  handleClick( event, event->worldIntersection() );
+}
+
+void Qgs3DMapToolMeasureLine::onTerrainEntityChanged()
+{
+  // no need to disconnect from the previous entity: it has been destroyed
+  // start listening to the new terrain entity
+  if ( QgsTerrainEntity *terrainEntity = mCanvas->scene()->terrainEntity() )
+  {
+    connect( terrainEntity->terrainPicker(), &Qt3DRender::QObjectPicker::clicked, this, &Qgs3DMapToolMeasureLine::onTerrainPicked );
+  }
+}
+
+void Qgs3DMapToolMeasureLine::handleClick( Qt3DRender::QPickEvent *event, const QgsVector3D &worldIntersection )
+{
+  if ( event->button() == Qt3DRender::QPickEvent::LeftButton )
+  {
+    if ( mDone )
+    {
+      mDialog->restart();
+    }
+
+    mDone = false;
+    QgsVector3D mapCoords = Qgs3DUtils::worldToMapCoordinates( QgsVector3D( worldIntersection.x(),
+                            worldIntersection.y(),
+                            worldIntersection.z() ), mCanvas->map()->origin() );
+    addPoint( QgsPoint( mapCoords.x(), mapCoords.y(), mapCoords.z() ) );
+    mDialog->show();
+  }
+  else if ( event->button() == Qt3DRender::QPickEvent::RightButton )
+  {
+    // Finish measurement
+    mDone = true;
+  }
+  else if ( event->button() == Qt3DRender::QPickEvent::MiddleButton )
+  {
+    undo();
+  }
+}
+
+void Qgs3DMapToolMeasureLine::setMeasurementLayerRenderer()
+{
+  mMeasurementLayer->setRenderer3D( mLineSymbolRenderer );
+  mLineSymbolRenderer->setLayer( mMeasurementLayer );
+}
+
+void Qgs3DMapToolMeasureLine::updateMeasurementLayer()
+{
+  double verticalScale = canvas()->map()->terrainVerticalScale();
+  QgsLineString *line;
+  if ( verticalScale != 1.0 )
+  {
+    QVector<QgsPoint> descaledPoints;
+    QVector<QgsPoint>::const_iterator it;
+    QgsPoint point;
+    for ( it = mPoints.constBegin(); it != mPoints.constEnd(); ++it )
+    {
+      point = *it;
+      descaledPoints.append(
+        QgsPoint( it->x(), it->y(), it->z() / verticalScale )
+      );
+    }
+    line = new QgsLineString( descaledPoints );
+  }
+  else
+  {
+    line = new QgsLineString( mPoints );
+  }
+  QgsGeometry *lineGeometry = new QgsGeometry( line );
```

Comment:
Also here you can just use a local variable `QgsGeometry lineGeometry( line );` - and avoid a memory leak. (the geometry map nor the changeGeometryValues() function do not take ownership of the passed geometry - they just make a copy)

Reply:
Done.